src/mlpack/tests/main_tests/lmnn_test.cpp

```diff
@@ -0,0 +1,936 @@
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_cols,
+      inputData.n_cols);
+}
+
+/**
+ * Ensure that when we pass optimizer of type lbfgs, we also get the desired
+ * shape of output.
+ */
+BOOST_AUTO_TEST_CASE(LMNNOptimizerTest)
+{
+  arma::mat inputData;
+  if (!data::Load("iris.csv", inputData))
+    BOOST_FAIL("Cannot load iris.csv!");
+
+  arma::Row<size_t> labels;
+  if (!data::Load("iris_labels.txt", labels))
+    BOOST_FAIL("Cannot load iris_labels.txt!");
+
+  // Input random data points.
+  SetInputParam("input", inputData);
+  SetInputParam("labels", labels);
+  SetInputParam("optimizer",  std::string("bbsgd"));
+
+  mlpackMain();
+
+  // Check that final output has expected number of rows and colums.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_cols,
+      inputData.n_cols);
+
+  // Reset rettings.
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  // Input random data points.
+  SetInputParam("input", inputData);
+  SetInputParam("labels", labels);
+  SetInputParam("optimizer",  std::string("sgd"));
+
+  mlpackMain();
+
+  // Check that final output has expected number of rows and colums.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_cols,
+      inputData.n_cols);
+
+  // Reset rettings.
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  // Input random data points.
+  SetInputParam("input", inputData);
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("optimizer",  std::string("lbfgs"));
+
+  mlpackMain();
+
+  // Check that final output has expected number of rows and colums.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("output").n_cols,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_rows,
+      inputData.n_rows);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("transformed_data").n_cols,
+      inputData.n_cols);
+}
+
+/**
+ * Ensure that when we pass a valid initial learning point, we get
+ * output of the same dimensions.
+ */
+BOOST_AUTO_TEST_CASE(LMNNValidDistanceTest)
```

Comment:
Ya, there is an LMNNInvalidDistanceTest as well just below it. I think that's what you're probably referring too.

Reply:
Sorry, If I not getting the point but I believe those are the exact same ideas valid and invalid distance tests are based on. Like valid one for 1st point and invalid one for 2nd point(checking square matrix). Please, correct me if I am off the point.

Reply:
Oh, sorry!  You are exactly right.  I had misread the valid distance test.  I thought that in this test, you were passing in a square matrix.

Do you think it would be worthwhile to then just add a third test to ensure that we can pass a distance matrix of size `inputData.n_rows` by `inputData.n_rows`?

Reply:
no worries! I will add it.

Reply:
Done!